pkg/action/executor/move_util.go

```diff
@@ -0,0 +1,363 @@
+		b1 := 0
+		if i < l1 {
+			if tmp, err := strconv.Atoi(a1[i]); err == nil {
+				b1 = tmp
+			}
+		}
+
+		b2 := 0
+		if i < l2 {
+			if tmp, err := strconv.Atoi(a2[i]); err == nil {
+				b2 = tmp
+			}
+		}
+
+		if b1 != b2 {
+			return b1 - b2
+		}
+	}
+
+	return 0
+}
+
+//retry to execute a function with a timeout
+func retryDuring(attempts int, timeout time.Duration, sleep time.Duration, myfunc func() error) error {
+	t0 := time.Now()
+
+	var err error
+	for i := 0; ; i++ {
+		if err = myfunc(); err == nil {
+			glog.V(4).Infof("[retry-%d/%d] success", i+1, attempts)
+			return nil
+		}
+
+		glog.V(4).Infof("[retry-%d/%d] Warning %v", i+1, attempts, err)
+		if i >= (attempts - 1) {
+			break
+		}
+
+		if timeout > 0 {
+			if delta := time.Now().Sub(t0); delta > timeout {
+				err = fmt.Errorf("after %d attepmts (during %v) last error: %v", i, delta, err)
+				glog.Error(err)
+				return err
+			}
+		}
+
+		if sleep > 0 {
+			time.Sleep(sleep)
+		}
+	}
+
+	err = fmt.Errorf("after %d attepmts last error: %v", attempts, err)
+	glog.Error(err)
+	return err
+}
+
+//TODO: check which fields should be copied
+func copyPodInfo(oldPod, newPod *api.Pod) {
+	//1. typeMeta
+	newPod.TypeMeta = oldPod.TypeMeta
+
+	//2. objectMeta
+	newPod.ObjectMeta = oldPod.ObjectMeta
+	newPod.SelfLink = ""
+	newPod.ResourceVersion = ""
+	newPod.Generation = 0
+	newPod.CreationTimestamp = metav1.Time{}
+	newPod.DeletionTimestamp = nil
+	newPod.DeletionGracePeriodSeconds = nil
+
+	//3. podSpec
+	spec := oldPod.Spec
+	spec.Hostname = ""
+	spec.Subdomain = ""
+	spec.NodeName = ""
+
+	newPod.Spec = spec
+	return
+}
+
+func calcGracePriod(pod *api.Pod) int64 {
+	grace := podDeletionGracePeriodDefault
+	if pod.Spec.TerminationGracePeriodSeconds != nil {
+		grace = *(pod.Spec.TerminationGracePeriodSeconds)
+		if grace > podDeletionGracePeriodMax {
+			grace = podDeletionGracePeriodMax
+		}
+	}
+	return grace
+}
+
+// move pod nameSpace/podName to node nodeName
+func movePod(client *kclient.Clientset, pod *api.Pod, nodeName string, retryNum int) (*api.Pod, error) {
+	podClient := client.CoreV1().Pods(pod.Namespace)
+	if podClient == nil {
+		err := fmt.Errorf("cannot get Pod client for nameSpace:%v", pod.Namespace)
+		glog.Error(err)
+		return nil, err
+	}
+
+	//1. copy the original pod
+	id := fmt.Sprintf("%v/%v", pod.Namespace, pod.Name)
+	glog.V(2).Infof("move-pod: begin to move %v from %v to %v",
+		id, pod.Spec.NodeName, nodeName)
+
+	npod := &api.Pod{}
+	copyPodInfo(pod, npod)
+	npod.Spec.NodeName = nodeName
+
+	//2. kill original pod
+	grace := calcGracePriod(pod)
+	delOption := &metav1.DeleteOptions{GracePeriodSeconds: &grace}
+	err := podClient.Delete(pod.Name, delOption)
+	if err != nil {
+		err = fmt.Errorf("move-failed: failed to delete original pod-%v: %v",
+			id, err)
+		glog.Error(err)
+		return nil, err
+	}
+
+	//3. create (and bind) the new Pod
+	time.Sleep(time.Duration(grace+1) * time.Second) //wait for the previous pod to be cleaned up.
```

Comment:
The original Pod is deleted, but it will only disappear after gracePeriod seconds.  We will create a new Pod with the same nameSpace/name as the original one. So if we don't wait for the original Pod to disappear, the creation for the new Pod will fail immediately.